docs/source/custom-collections.rst

```diff
@@ -0,0 +1,326 @@
+        After computation, the results will be returned in the same layout,
+        with the keys replaced with their corresponding outputs.
+
+
+.. staticmethod:: __dask_optimize__(dsk, keys, \*\*kwargs)
+
+    Given a graph and keys, return a new opimized graph.
+
+    This method can be either a ``staticmethod`` or a ``classmethod``, but not
+    an instancemethod.
+
+    If not implemented, defaults to returning the graph unchanged.
+
+    Parameters
+    ----------
+    dsk : MutableMapping
+    keys : list
+    \*\*kwargs
+        Extra keyword arguments forwarded from the call to ``compute`` or
+        ``persist``. Can be used or ignored as needed.
+
+    Returns
+    -------
+    optimized_dsk : MutableMapping
+        The optimized dask graph.
+
+
+.. staticmethod:: __dask_default_get__(dsk, keys, \*\*kwargs)
+
+    The default scheduler ``get`` to use for this object.
+
+    Usually attached to the class as a staticmethod, e.g.
+
+    >>> from dask.threaded import get as threaded_get
+    >>> class MyCollection(object):
+    ...     # Use the threaded scheduler by default
+    ...     __dask_default_get__ = staticmethod(threaded_get)
+
+
+.. method:: __dask_postcompute__(self)
+
+    Finalizer and (optional) extra arguments.
+
+    Used to implement ``dask.compute``.
+
+    Returns
+    -------
+    finalize : callable
+        A function with the signature ``finalize(results, *extra_args)``.
+        Called with the computed keys from ``__dask_keys__`` and any extra
+        arguments as specified in ``extra_args``. Should perform any necessary
+        finalization before returning from ``compute``. For example, the
+        ``finalize`` function for ``dask.array.Array`` concatenates all the
+        individual array chunks into one large numpy array, which is then the
+        result of ``compute``.
+    extra_args : tuple
+        Any extra arguments to pass to ``finalize`` after ``results``. If no
+        extra arguments should be an empty tuple.
+
+
+.. method:: __dask_postpersist__(self)
+
+    Rebuilder and (optional) extra arguments.
+
+    Used to implement ``dask.persist``.
+
+    Returns
+    -------
+    rebuild : callable
+        A function with the signature ``rebuild(dsk, *extra_args)``. Called
+        with a computed graph containing only the keys and results from
+        ``__dask_keys__``, as well as any extra arguments as specified in
+        ``extra_args``. Should return an equivalent dask collection with the
+        same keys as ``self``, but with the results already computed. For
+        example, the ``rebuild`` function for ``dask.array.Array`` is just the
+        ``__init__`` method called with the new graph but the same metadata.
+    extra_args : tuple
+        Any extra arguments to pass to ``rebuild`` after ``dsk``. If no extra
+        arguments should be an empty tuple.
+
+
+.. note:: It's also recommended to define ``__dask_tokenize__``,
+          see :ref:`deterministic-hashing`.
+
```

Comment:
It might be helpful to see all of these methods laid out in a mock `compute` function, just to get a context of how they are likely to be used.

Reply:
Strong 👍

Reply:
Added.

Reply:
We may eventually choose to rename this function from `postpersist` to something like `collection_from_graph`.  I can imagine it being used in other contexts.  I'm not suggesting any action now.  Just rambling.